### PR TITLE
[bugfix] Remove warning for distributed values

### LINKED ISSUE #7129: Training is stuck if ranks don't have the same keys in metrics dict at the end of validation epoch
## 🐛 Bug

In my D2Go model training, the [evaluation metrics are reduced on rank 0](https://github.com/facebookresearch/detectron2/blob/master/detectron2/evaluation/coco_evaluation.py#L151-L157) and logged in validation_epoch_end hook. The other ranks will just log an empty dict.
However, the recent change #6417 will always run _all_gather. I noticed that the training is stuck there until a NCCL timeout error kicks in. 

### To Reproduce

Use following [**BoringModel**](https://colab.research.google.com/drive/1HvWVVTK8j2Nj52qU4Q4YCyzOm0_aLQF3?usp=sharing) and post here

Apply follow code in `validation_epoch_end` hook and run with 2 GPUs.
```
def validation_epoch_end(self, outputs) -> None:
        rank = get_rank()
        res = {}
        if rank == 0:
            # reduce metric across ranks
            res = {"reduced_metric": 0.1}
        self.log_dict(res)
```
https://colab.research.google.com/drive/1HvWVVTK8j2Nj52qU4Q4YCyzOm0_aLQF3?usp=sharing
(note: I'm not sure how to run distributed training in colab, so I'm not able to reproduce in above notebook)

### Expected behavior
Do not try to reduce on missing metric keys.

### Environment

**Note**: `Bugs with code` are solved faster ! `Colab Notebook` should be made `public` !

* `IDE`: Please, use our python [bug_report_model.py](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pl_examples/bug_report_model.py
) template.

* `Colab Notebook`: Please copy and paste the output from our [environment collection script](https://raw.githubusercontent.com/PyTorchLightning/pytorch-lightning/master/tests/collect_env_details.py) (or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/PyTorchLightning/pytorch-lightning/master/tests/collect_env_details.py
# For security purposes, please check the contents of collect_env_details.py before running it.
python collect_env_details.py
```

 - PyTorch Version (e.g., 1.0):
 - OS (e.g., Linux):
 - How you installed PyTorch (`conda`, `pip`, source):
 - Build command you used (if compiling from source):
 - Python version:
 - CUDA/cuDNN version:
 - GPU models and configuration:
 - Any other relevant information:

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR remove the warning introduced in https://github.com/PyTorchLightning/pytorch-lightning/pull/6417/files.

Assuming users might want to log different keys on each rank, it becomes extremely complex to know if the key exists on each rank just to enable the warning as the Result objects won't exist in the EpochResultStore.

After going into the rabbit hole for 1 hour, decided remove the warning completely.

Fixes #7129 

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by awaelchli — APPROVED** [review]

**Review by ananthsub — APPROVED** [review]

File: pytorch_lightning/core/step_result.py [hunk 1]
```diff
@@ -576,7 +576,7 @@ def get_non_metrics_keys(self):
         """
         This function is used to filter metric keys for which the value isn't a Metric
         """
-        return [k for k, v in self.items() if not isinstance(v, Metric)]
+        return [k for k, v in self.items() if (not isinstance(v, Metric) and k != 'meta')]
```
Comment (ananthsub) [inline pytorch_lightning/core/step_result.py:579]: could you describe what changes now that `meta` is excluded?

Comment (tchaton) [reply pytorch_lightning/core/step_result.py:579]: The meta is an internal key `self['meta'] = {'_internal': {'_reduce_on_epoch': False, 'batch_sizes': []}}` and shouldn't be returned.

**Review by carmocca — APPROVED** [review]
Removed some more code which I believe you missed and fixed a test!

**Review by Borda — APPROVED** [review]
